src/socket.c

```diff
@@ -0,0 +1,221 @@
+/* vim: set tabstop=8 shiftwidth=4 softtabstop=4 expandtab smarttab colorcolumn=80: */
+/*
+ * Copyright (c) 2022 Nikos Mavrogiannopoulos
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <arpa/inet.h>
+#include <sys/select.h>
+#include <errno.h>
+#include <sys/wait.h>
+
+#include "socket.h"
+
+#define MAX(x,y) ((x)>(y)?(x):(y))
+
+typedef struct socket_list {
+	int s;
+	int family;
+	struct sockaddr addr;
+	socklen_t addrlen;
+	struct socket_list *next;
+} socket_list;
+
+static int listen_port(socket_list **slist, int port)
+{
+	struct addrinfo hints, *res, *ptr;
+	int y, r, s;
+	char portname[6], strip[64];
+	socket_list *lm;
+
+	snprintf(portname, sizeof(portname), "%d", port);
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_flags = AI_PASSIVE;
+
+	*slist = NULL;
+
+	/* listen to all available (IPv4 and IPv6) address */
+	if ((r = getaddrinfo(NULL, portname, &hints, &res)) != 0) {
+		fprintf(stderr, "getaddrinfo() failed: %s\n", gai_strerror(r));
+		return -1;
+	}
+
+	for (ptr = res; ptr != NULL; ptr = ptr->ai_next) {
+		s = socket(ptr->ai_family, SOCK_STREAM, 0);
+		if (s < 0) {
+			perror("socket() failed");
+			continue;
+		}
+
+		if (ptr->ai_family == AF_INET)
+			fprintf(stderr, "Listening on %s:%d\n", inet_ntop(ptr->ai_family,
+				&((struct sockaddr_in*)ptr->ai_addr)->sin_addr, strip,
+				sizeof(strip)), port);
+		else if (ptr->ai_family == AF_INET6)
+			fprintf(stderr, "Listening on [%s]:%d\n", inet_ntop(ptr->ai_family,
+				&((struct sockaddr_in6*)ptr->ai_addr)->sin6_addr, strip,
+				sizeof(strip)), port);
+
+#if defined(IPV6_V6ONLY)
+		if (ptr->ai_family == AF_INET6) {
+			y = 1;
```

Comment:
You are setting y = 1 some lines below so I guess this one is redundant

Reply:
It is a setsockopt option. I set the expected value of the option just before the call explicitly to avoid a rearrange of the lines messing what the actual setsockopt command does. I find that repeating cleaner when reading as well.